src/NUnitFramework/framework/Attributes/DataSourceAttribute.cs

```diff
@@ -0,0 +1,392 @@
+
+          // skip empty lines as denoted by a newline character
+          if (t == string.Empty || t == Environment.NewLine)
+            continue;
+          
+          // once we have a valid (non-empty) string I continue to process
+          var csv = t.Split(',');
+          
+          // check that dimensions are the same
+          if (csv.Length != paramDim)
+            throw new ArgumentException(String.Format("Method parameters ({0}) and number of columns ({1}) differ. They need to be equal.", paramDim, csv.Length));
+          
+          /*
+           * Based on the method signature, convert each column element respectively. For
+           * elements that have extra quotes (for encapsulation), we strip those out and include data (mainly strings)
+           * 
+           * This accounts for the following cases:
+           * 2,3,4
+           * "2","3","4"
+           * "a","b",c
+           */
+          
+          var plist = new ArrayList(csv.Length);
+          for (int i = 0; i < csv.Length; i++){
+            var paramType = paramInfo[i].ParameterType;
+            plist.Add(Convert.ChangeType(csv[i].Trim(new char[]{'\'', '"',' '}), paramType));
+          }
+          
+          p.Add(plist.ToArray());
+          k++;
+        }
+
+      }
+      
+      return p;
+    }
+    
+    /// <summary>
+    /// Helper function when iterating through a subset of rows
+    /// </summary>
+    /// <param name="i">current index</param>
+    /// <param name="N">upper bound</param>
+    /// <returns>TRUE if upper bound is zero or when the index is less than the upper bound. FALSE otherwise</returns>
+    /// <remarks>This corresponds to the <c>DataRowCount</c> enum where 0 represents all rows.</remarks>
+    private bool EndOfData(int i, int N){
+      
+      if(N == 0){ return true; }
+      if(i <= N) { return true; }
+
+      return false;
+    }
+    
+    
+    private IEnumerable GetXmlData(IMethodInfo method)
+    {
+
+      // get dimensions from method parameters
+      var paramInfo = method.GetParameters();
+      var paramDim = paramInfo.Length;
+      
+      // open file and check that dimensions match that of parameters
+      IList<object[]> pip = new List<object[]>();
+      
+      /* 
+       * Iterate through each row and generate the list of data parameters.
+       * Keep in mind that unlike a CSV file, the entire file is read into memory
+       * and it is properly formated. I use the defined functions of XmlDocument to make
+       * it easier to grab nodes via XPath. This is not memory efficient (like XmlReader)
+       */
+      
+      var xmlDoc = new XmlDocument();
+      xmlDoc.Load(SourceName);
+      
+      // select the root node and get child elements (rows)
+      var dataRows = xmlDoc.SelectSingleNode("//data").ChildNodes;
+      int N = DataRows != 0 ? DataRows : dataRows.Count;
+      
+      for (int i = 1; EndOfData(i,N);i++){
```

Comment:
Use of EndOfData in the loop seems to make the code somewhat obscure, especially since EndOfData returns true when we are NOT at the end of the data. Spelling out the test would make the for statement slightly more complex but probably easier to understand.